core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -339,7 +346,14 @@ void CudaSpace::impl_deallocate(
   }
 
   try {
+#ifndef CUDART_VERSION
+#no CUDART_VERSION
+#elif (CUDART_VERSION >= 11020)
+    CUDA_SAFE_CALL(cudaFreeAsync(arg_alloc_ptr, 0));
+    cudaDeviceSynchronize();
```

Comment:
I expect synchronization needs to happen before the deallocation.
```suggestion
    CUDA_SAFE_CALL(cudaDeviceSynchronize());
    CUDA_SAFE_CALL(cudaFreeAsync(arg_alloc_ptr, 0));
```

Reply:
Thanks for the comment, my thinking is that to replicate cudaFree using cudaFreeAsync and cudaDeviceSynchronize, the sync would go after the cudaFreeAsync call because if there is a use case that a user is rapidly allocating and deallocating, the sync would make sure that the allocation is in fact returned to the pool before the next allocation request. I don't believe there should be a sync before the cudaFreeAsync because there is a sync immediately right after the cudaDeviceSync, and as of right now we don't want to introduce any asynchronous behavior.

Reply:
Hi @dalg24, what are your thoughts on the comment above? I confirmed my suspicion with the CUDA team, and they actually mentioned that the sync was more or less functionally equivalent to before (or not at all, as long as the rest of the code does not have any loose async calls) as the memory is not freed until the prior GPU work completes, but there is a small perf bonus of including it after the call to free.

Reply:
@crtrott told me that memory allocations in Views are semantically fencing in Kokkos (at least as of now), so I think that for safety we should synchronize both before and after the free so that it's clear to the reader that no actually asynchronous workflows involving memory allocation are currently possible.

Reply:
Good call, I think that is a good add for readability, added and committed in https://github.com/kokkos/kokkos/pull/4026/commits/ac17a44cd61fabe88915db998118f8782e81e347

Reply:
so cudaFree and cudaMalloc do block all outstanding kernels right? Do cudaFreeAsync and cudaMallocAsync do so too? If not yes, you need to have that device syncrhonization before.